data/src/test/java/org/apache/iceberg/io/TestTaskDeltaWriter.java

```diff
@@ -0,0 +1,417 @@
+
+    WriteResult result = deltaWriter.complete();
+    commitTransaction(result);
+
+    Assert.assertEquals("Should have a data file.", 1, result.dataFiles().length);
+    Assert.assertEquals("Should have a pos-delete file", 1, result.deleteFiles().length);
+    DeleteFile posDeleteFile = result.deleteFiles()[0];
+    Assert.assertEquals("Should be a pos-delete file", FileContent.POSITION_DELETES, posDeleteFile.content());
+    Assert.assertEquals("Should have expected records", expectedRowSet(ImmutableList.of(
+        createRecord(4, "eee"),
+        createRecord(3, "fff"),
+        createRecord(2, "ggg"),
+        createRecord(1, "hhh")
+    )), actualRowSet("*"));
+
+    // Check records in the data file.
+    DataFile dataFile = result.dataFiles()[0];
+    Assert.assertEquals(ImmutableList.of(
+        createRecord(1, "aaa"),
+        createRecord(2, "bbb"),
+        createRecord(3, "ccc"),
+        createRecord(4, "ddd"),
+        createRecord(4, "eee"),
+        createRecord(3, "fff"),
+        createRecord(2, "ggg"),
+        createRecord(1, "hhh")
+    ), readRecordsAsList(table.schema(), dataFile.path()));
+
+    // Check records in the pos-delete file.
+    Schema posDeleteSchema = DeleteSchemaUtil.pathPosSchema();
+    Assert.assertEquals(ImmutableList.of(
+        posRecord.copy("file_path", dataFile.path(), "pos", 0L),
+        posRecord.copy("file_path", dataFile.path(), "pos", 1L),
+        posRecord.copy("file_path", dataFile.path(), "pos", 2L),
+        posRecord.copy("file_path", dataFile.path(), "pos", 3L)
+    ), readRecordsAsList(posDeleteSchema, posDeleteFile.path()));
+  }
+
+  @Test
+  public void testUpsertSameRow() throws IOException {
+    List<Integer> eqDeleteFieldIds = Lists.newArrayList(idFieldId, dataFieldId);
+    Schema eqDeleteSchema = table.schema();
+
+    GenericTaskDeltaWriter deltaWriter = createTaskWriter(eqDeleteFieldIds, eqDeleteSchema);
+
+    Record record = createRecord(1, "aaa");
+    deltaWriter.write(record);
+    deltaWriter.delete(record);
+    deltaWriter.write(record);
+    deltaWriter.delete(record);
+    deltaWriter.write(record);
+
+    WriteResult result = deltaWriter.complete();
+    Assert.assertEquals("Should have a data file.", 1, result.dataFiles().length);
+    Assert.assertEquals("Should have a pos-delete file and an eq-delete file", 2, result.deleteFiles().length);
+    commitTransaction(result);
+    Assert.assertEquals("Should have an expected record", expectedRowSet(ImmutableList.of(record)), actualRowSet("*"));
+
+    // Check records in the data file.
+    DataFile dataFile = result.dataFiles()[0];
+    Assert.assertEquals(ImmutableList.of(record, record, record), readRecordsAsList(table.schema(), dataFile.path()));
+
+    // Check records in the eq-delete file.
+    DeleteFile eqDeleteFile = result.deleteFiles()[0];
+    Assert.assertEquals(ImmutableList.of(record, record), readRecordsAsList(eqDeleteSchema, eqDeleteFile.path()));
```

Comment:
Is it necessary to encode the record twice? What about detecting already deleted keys and omitting the second delete? That might be over-complicating this for a very small optimization though.

Reply:
Em,  agreed that we don't have to upsert the same key twice.  The first upsert is enough.